torch/_functorch/aot_autograd.py

```diff
@@ -2772,6 +3038,277 @@ def traced_forward(*primals_fwd_seed_fwd_base_offset):
+    subclass_tangent_meta = []
+
+    metadata = ViewAndMutationMeta(
+        input_info=input_info,
+        requires_grad_info=requires_grad_info,
+        output_info=output_info,
+        num_intermediate_bases=num_intermediate_bases,
+        keep_input_mutations=keep_input_mutations,
+        traced_tangents=traced_tangents,
+        subclass_inp_meta=subclass_inp_meta,
+        subclass_out_meta=subclass_out_meta,
+        subclass_tangent_meta=subclass_tangent_meta,
+    )
+    return metadata
+
+# Given a function operating on Subclass -> Subclass, returns an function that operates on Tensor -> Tensor
+# Also returns:
+# - the new set of arguments to pass into this function (now that tensor subclasses have been eliminated)
+# - the updated ViewAndMutationMeta for this dense -> dense function.
+# The other important arguments are:
+# - flat_fn_maybe_joint: when trace_joint=True, this is the joint fw-bw function.
+#                        when trace_joint=False, this is just the forward function.
+# - fw_only: this is *always* the forward-only function.
+#   Why do we need this? We need to collect updated ViewAndMutationMeta on our new dense -> dense functions.
+#   In particular, we need this to tell the partitioner how many dense forward outputs there are.
+def aot_dispatch_subclass(
+    flat_fn_maybe_joint,
+    args: List[Any],
+    *,
+    trace_joint: bool,
+    meta: ViewAndMutationMeta,
+    fw_only: Callable,
+) -> Tuple[Callable, List[Any], Optional[SubclassMeta]]:
+    # Skip logic if we don't need to trace through any subclasses
+    req_subclass_dispatch = requires_subclass_dispatch(args, meta)
+    if not req_subclass_dispatch:
+        return flat_fn_maybe_joint, args, None
+
+    # TODO: add subclass guards (later PR).
+
+    # What's going on here? We need to compute subclass metadata about the outputs of the joint (grad_inputs).
+    # Annoying: we don't know the grad input metas until we're in the middle of tracing the joint,
+    # so we set it later, while we're tracing the joint (see inner_fn() below).
+    # Another option would be to run our run_functionalized_fw_and_collect_metadata() function
+    # directly on the joint, but this would hurt compile time (adding yet another pass through the joint).
+    subclass_meta = SubclassMeta()
+
+    def inner_fn(fn, args, *, use_trace_joint: bool):
+        # Step 1: wrap tensor inputs into subclasses if necessary
+        all_args = wrap_tensor_subclasses(args, trace_joint=use_trace_joint, meta=meta)
+
+        # Step 2: call the inner function, with our (maybe subclass) inputs
+        wrapped_outs = fn(*all_args)
+
+        if use_trace_joint:
+            # See Note: [Computing Subclass Metadata about grad_inputs]
+            # We also stash subclass info on our grad_inputs, if we're tracing the joint.
+            nonlocal subclass_meta
+            assert isinstance(wrapped_outs, tuple) and len(wrapped_outs) == 2
+            grad_inputs = wrapped_outs[1]
```

Comment:
don't need the forward outs, cuz we already knew what their metas were from the forward analysis